src/borrower/feed/navfeed.sol

```diff
@@ -273,106 +161,272 @@ contract NAVFeed is BaseNFTFeed, Interest, Buckets, FixedPoint {
-    function _calcTotalDiscount() internal view returns(uint, uint) {
-        uint normalizedBlockTimestamp = uniqueDayTimestamp(block.timestamp);
-        uint sum = 0;
+    // --- NAV calculation ---
+    function currentNAV() public view returns(uint) {
+        uint nnow = uniqueDayTimestamp(block.timestamp);
+        uint nLastUpdate = uniqueDayTimestamp(lastNAVUpdate);
+        uint overdueButNotWrittenOff_ = 0;
+        for (uint i = nLastUpdate; i < nnow; i = i + 1 days) {
+            overdueButNotWrittenOff_ = safeAdd(overdueButNotWrittenOff_, buckets[i]);
+        }
 
-        uint currDate = discountStartPointer;
+        return safeAdd(safeAdd(currentDiscount(), safeAdd(overdueButNotWrittenOff, overdueButNotWrittenOff_)), currentWriteOffs());
+    }
 
-        if (currDate > lastBucket  || discountStartPointer == 0) {
-            return (0, discountStartPointer);
+    function currentDiscount() public view returns(uint) {
+        if (latestDiscount == 0) {
+            return 0;
         }
 
-        while(currDate != NullDate && currDate < normalizedBlockTimestamp) {
-            currDate = buckets[currDate].next;
+        uint nnow = uniqueDayTimestamp(block.timestamp);
+        uint nLastUpdate = uniqueDayTimestamp(lastNAVUpdate);
+
+        uint totalDiscount = rmul(latestDiscount, rpow(discountRate.value, safeSub(nnow, nLastUpdate), ONE));
+
+        uint diff = 0;
+        for(uint i = nLastUpdate; i < nnow; i = i + 1 days) {
+            // nLastUpdate < i < nnow
+            // 2 days ago < 1 day ago < now
+            diff = safeAdd(diff, rmul(buckets[i], rpow(discountRate.value, safeSub(nnow, i), ONE)));
         }
 
-        // next time we can start here to find the next first bucket in the future
-        uint nextStartPointer = currDate;
+        totalDiscount = secureSub(totalDiscount, diff);
 
-        while(currDate != NullDate)
-        {
-            sum = safeAdd(sum, calcDiscount(buckets[currDate].value, normalizedBlockTimestamp, currDate));
-            currDate = buckets[currDate].next;
+        // TODO: fix rounding errors that this if statement is not required anymore
+        if (totalDiscount == 1) {
+            return 0;
         }
-        return (sum, nextStartPointer);
-    }
 
-    function currentNAV() public view returns(uint) {
-        (uint nav, ) = _currentNAV();
-        return nav;
+        return totalDiscount;
     }
 
-    // returns the NAV (net asset value) of the pool
-    function _currentNAV() internal view returns(uint, uint) {
-        // calculates the NAV for ongoing loans with a maturityDate date in the future
-        (uint nav_, uint nextStartPointer) = _calcTotalDiscount();
-        // include ovedue assets to the current NAV calculation
-        for (uint i = 0; i < writeOffs.length; i++) {
+    function currentWriteOffs() public view returns(uint) {
+        uint sum = 0;
+        for (uint i = 0; i < writeOffGroups.length; i++) {
```

Comment:
I think we could safe some gas if the loop would start at `WRITEOFF_RATE_GROUP_START`

Reply:
The writeoff groups don't start at `WRITEOFF_RATE_GROUP_START`. Only the rate groups in the pile that correspond to these writeoff groups start at `WRITEOFF_RATE_GROUP_START`.

Reply:
`WRITEOFF_RATE_GROUP_START` needs a better name :)

Reply:
Ahh, I missed the `writeOffGroups[i].percentage.value` part.